### PR TITLE
Added MB_X_Y_STAIR_WARP metatile behaviors

### PR DESCRIPTION
**Pull request by pkmnsnfrn**
INCORPORATES A FEATURE BRANCH - DO NOT SQUASH!

# Description
Player navigates around sideways stairs

* Ported @ghoulslash's [`stair_warps`](https://github.com/ghoulslash/pokeemerald/tree/stair_warps) branch to expansion

# Usage

## Metatile Behaviors
No tilesets within vanilla Emerald have been modified. Users will need to manually add the new metatile behaviors to the relevent metatiles.

## Functionality
When player is standing on a metatile with `MB_X_Y_STAIR_WARP` behavior and walks towards a warp, they player's x and y position changes to simulate the player going up and down the stairs.

# Testing
## Clean Branch
You can recreate this branch by applying a patch or pulling the repo. From a clean version of expansion's upcoming, you can either:

### Patch
`wget https://files.catbox.moe/3o61ki.patch -O stair_warps.patch ; git apply stair_warps.patch ; rm stair_warps.patch`

### Repo
`git remote add psf-expansion https://github.com/PokemonSanFran/pokeemerald-expansion/ ; git pull psf-expansion stair_warps`

## Manual Tests
After replicating the branch, to recreate my testing environment, you can either directly download the debug script and config file, or manually create the changes.

### Download
`wget https://github.com/rh-hideout/pokeemerald-expansion/commit/6ea73bb8b41615deaab776b0ffac11061228a5b6.patch -O stair_warps_testing.patch ; git apply stair_warps_testing.patch ; rm stair_warps_testing.patch ; wget https://github.com/rh-hideout/pokeemerald-expansion/commit/e44beb77fd2b34f0b4db4789b5d0ecbfcf156cf8.patch -O stair_warps_testing2.patch ; git apply stair_warps_testing2.patch ; rm stair_warps_testing2.patch`

### Manual Testing
* Start a New Game
* R + Start to open Overworld Debug Menu
* Utilites > Cheat Start
* R + Start to open Overworld Debug Menu
* Utilities > Warp to map warp > Group 24 > Map 7 > Warp 0
* Walk up and down every stair case
* Run up and down every stair case
* Use the Acro Bike up and down every stair case

## Verified Scenarios
The video shows all the Manual Testing.

https://github.com/user-attachments/assets/63b4cea6-95f0-4a52-89c7-15aa5f097210

# **People who collaborated with me in this PR**
This was originally [written by](<https://github.com/ghoulslash/pokeemerald/tree/stair_warps>) by @ghoulslash.

# Discord Contact Info
I am `pkmnsnfrn` on Discord.

### CODE REVIEW

**Comment by AlexOn1ine** [comment]
I assume this is the same pr as  #4836?

**Comment by pkmnsnfrn** [comment]
> I assume this is the same pr as #4836?

incorrect.

**Comment by Bassoonian** [comment]
Please solve conflicts

File: src/field_screen_effect.c [hunk 1]
```diff
@@ -1388,3 +1400,237 @@ void FieldCB_RushInjuredPokemonToCenter(void)
...
+#undef tSpeedX
+#undef tSpeedY
+#undef tOffsetX
+#undef tOffsetY
+#undef tTimer
+#undef tDelay
+
+bool32 IsDirectionalStairWarpMetatileBehavior(u16 metatileBehavior, u8 playerDirection)
+{
+    switch (playerDirection)
```
Comment (Bassoonian) [inline src/field_screen_effect.c:1620]: I personally wonder if this is worth a switch case, but that's subjective. If so: indentation

File: src/field_screen_effect.c [hunk 2]
```diff
@@ -1304,21 +1316,21 @@ static bool32 PrintWhiteOutRecoveryMessage(u8 taskId, const u8 *text, u32 x, u32
...
+            gTasks[taskId].tPrintState = 1;
+            break;
+        case 1:
+            RunTextPrinters();
+            if (!IsTextPrinterActive(windowId))
+            {
+                gTasks[taskId].tPrintState = 0;
+                return TRUE;
+            }
+            break;
```
[the remark this answers is not in the document]
Comment (pkmnsnfrn) [reply src/field_screen_effect.c]: my IDE re-indented the file, I'll just revert this.

[cf2f85bdd7](https://github.com/rh-hideout/pokeemerald-expansion/pull/5278/commits/cf2f85bdd71588b0c6fc25b299295b3c87edfd64)

File: src/field_screen_effect.c [hunk 3]
```diff
@@ -1388,3 +1400,237 @@ void FieldCB_RushInjuredPokemonToCenter(void)
...
+            Overworld_PlaySpecialMapMusic();
+            WarpFadeInScreen();
+            LockPlayerFieldControls();
+            ExitStairsMovement(&tSpeedX, &tSpeedY, &tOffsetX, &tOffsetY, &tTimer);
+            tState++;
+            break;
+        case 1:
+            if (!WaitStairExitMovementFinished(&tSpeedX, &tSpeedY, &tOffsetX, &tOffsetY, &tTimer))
+                tState++;
+            break;
```
[the remark this answers is not in the document]
Comment (pkmnsnfrn) [reply src/field_screen_effect.c]: [cf2f85bdd7](https://github.com/rh-hideout/pokeemerald-expansion/pull/5278/commits/cf2f85bdd71588b0c6fc25b299295b3c87edfd64)

File: src/field_screen_effect.c [hunk 4]
```diff
@@ -1388,3 +1400,237 @@ void FieldCB_RushInjuredPokemonToCenter(void)
...
+            UpdateStairsMovement(tSpeedX, tSpeedY, &tOffsetX, &tOffsetY, &tTimer);
+            if (!PaletteFadeActive() && BGMusicStopped())
+                tState++;
+            break;
+        default:
+            gFieldCallback = FieldCB_DefaultWarpExit;
+            WarpIntoMap();
+            SetMainCallback2(CB2_LoadMap);
+            DestroyTask(taskId);
+            break;
```
[the remark this answers is not in the document]
Comment (pkmnsnfrn) [reply src/field_screen_effect.c]: [cf2f85bdd7](https://github.com/rh-hideout/pokeemerald-expansion/pull/5278/commits/cf2f85bdd71588b0c6fc25b299295b3c87edfd64)

File: src/field_screen_effect.c [hunk 5]
```diff
@@ -1388,3 +1400,237 @@ void FieldCB_RushInjuredPokemonToCenter(void)
...
+            LockPlayerFieldControls();
+            FreezeObjectEvents();
+            CameraObjectFreeze();
+            tState++;
+            break;
+        case 1:
+            if (!ObjectEventIsMovementOverridden(playerObjectEvent) || ObjectEventClearHeldMovementIfFinished(playerObjectEvent))
+            {
+                if (tDelay != 0)
+                    tDelay--;
```
[the remark this answers is not in the document]
Comment (pkmnsnfrn) [reply src/field_screen_effect.c]: [eb1469af1d](https://github.com/rh-hideout/pokeemerald-expansion/pull/5278/commits/eb1469af1dfd0c0c83aab862a6322bd44256d17e)

File: src/field_screen_effect.c (hunk 1 above)
Comment (pkmnsnfrn) [reply src/field_screen_effect.c]: [`874d4d0` (#5278)](https://github.com/rh-hideout/pokeemerald-expansion/pull/5278/commits/874d4d047002a7651a4767e145c8b71a98996a84)

File: src/metatile_behavior.c [hunk 6]
```diff
@@ -1404,17 +1404,60 @@ bool8 MetatileBehavior_IsTrainerHillTimer(u8 metatileBehavior)
         return FALSE;
 }
 
-bool32 MetatileBehavior_IsSignpost(u32 metatileBehavior)
+bool8 MetatileBehavior_IsDirectionalUpRightStairWarp(u8 metatileBehavior)
+{
+    if(metatileBehavior == MB_UP_RIGHT_STAIR_WARP)
```
Comment (Bassoonian) [inline src/metatile_behavior.c:1409]:
```suggestion
    if (metatileBehavior == MB_UP_RIGHT_STAIR_WARP)
```

**Review by Bassoonian — APPROVED** [review]

**Pull request merged**